How to set up new model documentation

jacobcook1995 opened the pull request:
# Description

This is my first attempt at a how to doc for setting up a new model, let me know if it makes sense and more importantly whether it actually covers all the necessary steps.

Fixes #148 

## Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Documentation improvement 

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

jacobcook1995 commented on docs/source/development/defining_new_models.md:
````diff
@@ -0,0 +1,221 @@
+touch virtual_rainforest/models/freshwater/model.py
+```
+
+This script must import the [`BaseModel` class](../api/core/model.md) as new class must
````
Just to mention I am aware that this link is broken, it links to a page that exists in the `feature/data_system` branch but not in `develop`, so once that is merged the link won't be broken any longer.

davidorme commented on docs/source/development/defining_new_models.md:
```diff
@@ -0,0 +1,221 @@
+
+The new model class is created using a class method, this means that a model name must
+be provided when class inheritance occurs, so that the model can automatically be added
+to the registry under that name. This name does not strictly have to match the
+`self.name` attribute, but it generally should.
```
Maybe the registry should extract `cls.name` rather than having duplicate locations - there is no reason why these shouldn't be identical?

jacobcook1995 replied:
Yeah relatively easy fix. I've added the change to the pull request

davidorme replied:
Honestly, I'd do it the other way around. So rather than synchronising two locations:

```python
    @classmethod
    def __init_subclass__(cls):
        """Method that adds new model classes to the model registry."""

        if getattr(cls, 'name') is None:
             LOGGER.critical("Model is missing the model name class attribute")
        if not isinstance(cls.name, str)
             LOGGER.critical("Model name is not a string")

        # Add the grid type to the registry
        if cls.name in MODEL_REGISTRY:
            LOGGER.warning(
                "Model with name %s already exists and is being replaced", cls.name 
            )
        MODEL_REGISTRY[cls.name ] = cls
```

davidorme replied:
I think there's a way of decorating a class attribute so that is part of the ABC and hence cannot be left empty. Still might have bogus content though.

davidorme replied:
And maybe use `f"The 'name' class attribute is not set in model class {cls.__name__}"`

jacobcook1995 replied:
I kind of prefer supplying a model name though, think it makes it clearer that a something other than normal class inheritance is being used.

jacobcook1995 replied:
I'm only keeping the `self.name` in `BaseModel` so that the base model unit tests don't break, could just remove that name entirely, and rewrite the unit tests a bit

davidorme replied:
The problem though, is that they should be identical, and I think it is _much_ stranger to have a class attribute written in the code that is then replaced with another value.

davidorme replied:
I think `cls.name` is important. When the model is created and running, that is the only place where the model name is stored and hence can be used as `self.name` in logger calls etc. And that name is _fundamentally_ an attribute of the class.

davidorme replied:
Also it isn't great to have a class attribute that doesn't appear in the code with a docstring. So, it should be there in the subclass definitions.

jacobcook1995 replied:
I think I might be misunderstanding something here. Are `cls.name` and `self.name` different things?

jacobcook1995 replied:
I was imagining that `cls.name` was `self.name` for the class in question

davidorme replied:
They do refer to the same thing, but in the context of the `class` method, which uses `cls` as an argument to indicate that it is referring to the properties of the class itself, rather than `self.name` referring to the properties of an _instance_ of the class. The point of class attributes specifically is that property value is set across all possible instances of the class.

davidorme replied:
But in defining a subclass, which has a class attribute `name` (maybe we should using `model_name`), I think that attribute should be set and docstring in the model code, rather than set via an argument to `__init_subclass`. It won't appear in the class attributes for example. Given that needs to happen, it makes sense to use it to register the class?

jacobcook1995 replied:
So the issue with the current way I'm doing it is that only class instances produced using `__init_subclass__` would end with a name populated?

jacobcook1995 replied:
If someone called `SoilModel(**kwargs)`, it would create an instance without a name?

jacobcook1995 replied:
Or at the moment I think it would be called "base", but was planning on deleting that

davidorme replied:
The use of `class ModelName(BaseModel)` _guarantees_ that `BaseModel.__init_subclass__` will be called, so `cls.name` would be populated on any BaseModel subclass. **But** that is an obscure way to populate a class attribute and those are typically included and documented as:

```python
class ModelName(BaseModel):

     model_name = "ModelName"
     """The model name for use in registering the model and logging"
``

davidorme replied:
No - the name is _only_ set in the creation of the class itself. So your `soil.model` object _creates_ the `SoilModel` object, and at that point the subclass itself (not a class instance) is registered using `__init_subclass__`.

If someone creates an _instance_ using `SoilModel(**kwargs)`, then that creates an instance, with the `name` class attribute populated when the package was loaded. The difference here is really where it is clearer to set that attribute, and I think the direct option above is better for clarity and having attributes appear in rendered docstrings.

jacobcook1995 replied:
Happy to switch it around if you think that makes sense! My only hesitation is that it feels like a fairly big functional change to the code (which would require new tests, etc). Wonder if that is sensible to be adding to a pull request about docs, or whether I should make a new issue + PR for it?

davidorme replied:
@dalonsoa @alexdewar - Thoughts? It's not a large change but does need code change in the BaseModel and tests. Seems like a small PR to me. And also - am I talking nonsense?

alexdewar replied:
@davidorme Sounds sensible to me.

I would do it more simply though (I didn't know `__init_subclass__` was a thing till now). I'd just have a check in `BaseModel`'s `__init__` along the lines of:

```py
if not hasattr(self, "name"):
    raise ValueError("Models must have a name attribute")
````

vgro commented on docs/source/development/defining_new_models.md:
```diff
@@ -0,0 +1,216 @@
+The Virtual Rainforest is designed to be modular, meaning that the set of models to be
+used in a particular run is configurable at the start of the simulation. We are starting
+out by defining a core set of models (`abiotic`, `animals`, `plants` and `soil`), which
+will generally all be included for the vast majority of simulations. In future, it might
```
```suggestion
will generally all be included for the vast majority of simulations. In the future, it might
```

vgro commented on docs/source/development/defining_new_models.md:
```diff
@@ -0,0 +1,216 @@
+used in a particular run is configurable at the start of the simulation. We are starting
+out by defining a core set of models (`abiotic`, `animals`, `plants` and `soil`), which
+will generally all be included for the vast majority of simulations. In future, it might
+be desirable to include models for other aspects of rainforests (e.g. `freshwater`), or
```
```suggestion
be desirable to include models for other aspects of rainforests (e.g. `freshwater`, `disturbance`), or
```

vgro replied:
I think it might be worth mentioning disturbance as this will be one of the major applications

vgro commented on docs/source/development/defining_new_models.md:
```diff
@@ -0,0 +1,216 @@
+
+The new model class is created using a class method, this means that a model name must
+be provided when class inheritance occurs, so that the model can automatically be added
+to the registry under that name. N.B. The `self.name` attribute for the class is
```
What does N.B. mean?

jacobcook1995 replied:
I meant it to be nota bene, but realise that you are supposed to either abbreviate that as "n.b." or "NB". Looking at it again though, I think it's a bad way to start a sentence so I will switch it to "It should be noted" or similar

vgro commented on docs/source/development/defining_new_models.md:
```diff
@@ -0,0 +1,216 @@
+    invalid rather than returning an initialised model instance an error is raised.
+
+    Args:
+        config: The complete (and validated) virtual rainforest configuration.
```
General comment/question: do we capitalize 'Virtual Rainforest' or 'virtual rainforest'

davidorme replied:
I think "Virtual Rainforest" in text? @robewers01

jacobcook1995 replied:
I've been capitalising it generally (though I didn't here 😅), but agree that should probably settle on an agreed style

vgro commented on docs/source/development/defining_new_models.md:
```diff
@@ -0,0 +1,216 @@
+
+## Ensuring that schema and models are always added to the registry
+
+At the moment,a configuration schema only get added to the schema registry when the
```
```suggestion
At the moment, a configuration schema only gets added to the schema registry when the
```

vgro reviewed (commented):
This is a very clear step by step description, very easy to follow. I'll try it out and let you know if something doesn't work or is missing. I made a few minor comments.

vgro commented:
Could you please describe all the imports that are necessary at the start, for example the numpy stuff and the required logger component? I found it in the sol example but not here in the documentation. Maybe a complete example file would be good.

davidorme commented on docs/source/development/defining_new_models.md:
````diff
@@ -0,0 +1,233 @@
+```python
+# One of the member functions of the Model class returns a class instance. mypy doesn't
+# know how to handle this unless annotations are imported from __future__
+from __future__ import annotations
+
+# Any needed for type hints of the config dictionary as the values are of various types
````
I believe you can simply quote the return name to avoid the `mypy` error too. We should do this systematically though.

jacobcook1995 replied:
Didn't realise that worked! Prefer this way, partly because its a little neater, partly because I like importing from `__future__`

vgro reviewed (approved):
Documentation looks good to me, the technical details I cannot judge at the moment.

alexdewar reviewed (commented):
Sorry, added myself as a reviewer by mistake. Looks good though :laughing:

davidorme commented:
Oddly can't reply direct to that comment @alexdewar. Don't know why!

We need `__init_subclass__` (it was new to me too!) to register the model when the package loads anyway. Does it make makes sense to keep the tests here to make it fall over before before someone actually tries to make an instance of a nameless model, hence calling `super().__init__()`

jacobcook1995 commented:
I've moved where the `model_name`'s get set to be defined in each model and then pulled from there to be used as a registry name. I put the checks in the `__init_subclass__` rather than the `__init__` so that invalid models cannot be added to the registry.

I also had to bump the version of `isort` as the `pre-commit` check had started failing (on Github) with the old version. I updated the minimum value in `pyproject.toml` to match this + generated a new poetry lock

davidorme commented:
I've just PR'd that patch on `isort` on `develop` to merge in #122 . Shouldn't be an issue.
